### ISSUE TITLE
Expression evaluation returns true (instead of an error) for "a is X" when X is an unresolved identifier

### ISSUE THREAD

**Issue by DanTup** [flutter/flutter#157922]
### Steps to reproduce

Moving from https://github.com/dart-lang/sdk/issues/56300 based on @bkonyi's comments.

---

@FMorschel reported to me conditional breakpoints not working as expected, and we tracked it back to expression evaluation returning `true` for conditions that we didn't expect. We managed to reproduce it with just simple expression eval:

![image](https://github.com/user-attachments/assets/f1c5e5d0-ec68-4119-938d-00cf188eb64f)

The error messages shown here only show up because I'm running in `-verbose` mode, otherwise you'd only see the `true` responses. When doing `x is A` where `A` is some class that's apparently not in scope, the result is just `true`. For non-Flutter dart, you get a reasonable error:

![image](https://github.com/user-attachments/assets/5b8722c6-5d1d-4973-a01e-2d28c64d4a6d)

The VM traffic is below. The `flutter run` `stdout` output shows the error in verbose mode, but the VM Service response to `evaluateInFrame` returns a valid boolean of `true` with no indication that the request failed.

### Expected results

There should be an error message, like in Dart, so it's clear the condition failed and didn't evaluate to `true`.

### Actual results

The expression returns as if it evaluated to `true`.

### Code sample

<details open><summary>Code sample</summary>

```dart
// Evaluate this in the Debug Console while running a Flutter app
a is SomeClassThatDoesNotExist
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

![image](https://github.com/user-attachments/assets/8963a107-1798-45aa-9a11-6f28224bcfc2)

![image](https://github.com/user-attachments/assets/a7d26452-adcd-4ca2-a258-0067e1c4eb7f)

</details>

### Logs

<details open><summary>Logs</summary>

```console
==> [VM] {
	"jsonrpc": "2.0",
	"id": "741",
	"method": "evaluateInFrame",
	"params": {
		"isolateId": "isolates/3676968675006663",
		"frameIndex": 0,
		"expression": "a is FutureOr<int>",
		"disableBreakpoints": true
	}
}
<== [Flutter] [+44567 ms] org-dartlang-debug:synthetic_debug_expression:1:6: Error: 'FutureOr' isn't a type.
<== [Flutter] [        ] a is FutureOr<int>
<== [Flutter] [        ]      ^^^^^^^^
<== [VM] {
	"jsonrpc": "2.0",
	"result": {
		"type": "@Instance",
		"_vmType": "Bool",
		"class": {
			"type": "@Class",
			"fixedId": true,
			"id": "classes/62",
			"name": "bool",
			"location": {
				"type": "SourceLocation",
				"script": {
					"type": "@Script",
					"fixedId": true,
					"id": "libraries/@0150898/scripts/dart%3Acore%2Fbool.dart/0",
					"uri": "dart:core/bool.dart",
					"_kind": "kernel"
				},
				"tokenPos": 508,
				"endTokenPos": 9302,
				"line": 15,
				"column": 1
			},
			"library": {
				"type": "@Library",
				"fixedId": true,
				"id": "libraries/@0150898",
				"name": "dart.core",
				"uri": "dart:core"
			}
		},
		"identityHashCode": 0,
		"kind": "Bool",
		"fixedId": true,
		"id": "objects/bool-true",
		"valueAsString": "true"
	},
	"id": "741"
}
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[√] Flutter (Channel master, 3.27.0-1.0.pre.51, on Microsoft Windows [Version 10.0.22631.4317], locale en-GB)
    • Flutter version 3.27.0-1.0.pre.51 on channel master at D:\Dev\Google\Flutter\Flutter main
    • Upstream repository git@github.com:DanTup/flutter.git
    • FLUTTER_GIT_URL = git@github.com:DanTup/flutter.git
    • Framework revision 6829c6a9b9 (2 weeks ago), 2024-10-15 22:36:36 +0800
    • Engine revision 107ea8baa4
    • Dart version 3.6.0 (build 3.6.0-334.2.beta)
    • DevTools version 2.40.1

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 35.0.0-rc3)
    • Android SDK at C:\Users\danny\AppData\Local\Android\sdk
    • Platform android-35, build-tools 35.0.0-rc3
    • Java binary at: C:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0--11185874)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.10.5)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.10.35122.118
    • Windows 10 SDK version 10.0.22621.0

[√] Android Studio (version 2023.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.9+0--11185874)

[√] VS Code (version 1.95.0)
    • VS Code at C:\Users\danny\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.101.20241031

[√] VS Code (version 1.93.0-insider)
    • VS Code at C:\Users\danny\AppData\Local\Programs\Microsoft VS Code Insiders
    • Flutter extension version 3.95.20240801

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22631.4317]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 130.0.6723.91
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 127.0.2651.86

[√] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by huycozy** [comment]
I can observe the issue on Flutter stable `3.24.4` as well. Marking this as a tool issue based on https://github.com/dart-lang/sdk/issues/56300#issuecomment-2248866797, please correct if it's wrong.

**Comment by bkonyi** [comment]
It looks like there's a couple of things that need to be fixed here.

First of all, the frontend server's `compileExpression` implementation doesn't actually clear the compiler's set of errors before compiling an expression, meaning that the errors count isn't reset and can't be relied on to detect if there's new errors. I have a fix up for review [here](https://dart-review.googlesource.com/c/sdk/+/394220).

Second, the Flutter tool's logic assumes that expression compilation requests are successful as long as the response includes kernel data. This isn't the case. For example, here's some human readable kernel output from the frontend server after evaluating `is A`:

```
main = <No Member>;
library from "evaluate:source" as sou {

  class #DebugClass extends fra::StatelessWidget {
    method /* from org-dartlang-debug:synthetic_debug_expression */ :Eval(fra::StatelessElement context, gal::GalleryOptions options) → dynamic
      return invalid-expression "org-dartlang-debug:synthetic_debug_expression:1:1: Error: This couldn't be parsed.\nis A\n^" is invalid-type;
  }
}
```

Once the CL above lands and rolls into Flutter, we need to check the error count and return a expression compilation RPC error.
